Enable ILASM for *nix

kyulee1 opened the pull request:
This enables ILASM for x-platforms.
1. The prebuilt file asmparse.c is renamed to asmparse.cpp -- clang doesn't support treating C file as C++.
2. Added a bunch of warning disable options mostly due to this prebuilt asmparse.cpp
3. Create a separte entry point "main" to pass WCHAR arguments.
4. PDB (CorSymWriter) part is disabled.
5. Converting/embedding resource file to binary is disabled.
6. jkotas kindly provided a code for IsTextUnicode which is unavailable on CoreCLR.

kyulee1 commented:
@janvorli @jkotas Can you please take a look?

benpye commented:
Would it make sense to add IsTextUnicode to PAL instead?

kangaroo commented:
Also, clang completely supports compiling a .c as a .cpp fwiw.  -x c++ is your friend.

janvorli commented on src/dlls/mscorpe/ceefilegenwriter.cpp:
```diff
@@ -420,13 +427,13 @@ HRESULT CeeFileGenWriter::generateImage(void **ppImage)
     if (! outputFileName && ppImage == NULL) {
         if (m_comImageFlags & COMIMAGE_FLAGS_IL_LIBRARY)
-            outputFileName = L"output.ill";
+            outputFileName = outputFileNameIll;
```
Could you please fix the write method to accept const string instead as the filename? There is no reason for it to modify it.

janvorli replied:
Btw, the problem is that the write method parameter uses "const LPWSTR" type, which is not what it really wanted. const LPWSTR makes the string pointer constant, not the string characters themselves. LPCWSTR would do the trick.

janvorli commented on src/dlls/mscorpe/pewriter.cpp:
```diff
@@ -2026,7 +2028,11 @@ HRESULT PEWriter::fixup(CeeGenTokenMapper *pMapper)
                             IMAGE_SECTION_HEADER* phdr = textSection->m_header;
                             // Add to reloc table
+#ifdef FEATURE_PAL
+                            ir.u.VirtualAddress = VAL32(rva);
```
Could you please rather change the IMAGE_RELOCATION to use anonymous union? Then you can use the same macro as for Windows. Anonymous unions are now allowed both in C and C++ (in C starting from C11 which we set in the compiler options)

janvorli commented on src/ilasm/CMakeLists.txt:
```diff
@@ -17,33 +18,53 @@ set(ILASM_SOURCES
-  ${ILASM_SOURCES}
-)
+if(CLR_CMAKE_PLATFORM_UNIX)
+  add_compile_options(-Wno-delete-non-virtual-dtor)
```
Could you please create a PR issue to fix these warnings or fix them if there are just a few of them?

janvorli commented on src/ilasm/main.cpp:
```diff
@@ -66,7 +69,7 @@ void MakeProperSourceFileName(__in __nullterminated WCHAR* wzOrigName,
         if((wzProperName[j] == '\\')||(j == 0))
         {
-            wcscat_s(wzProperName,MAX_FILENAME_LENGTH,L".il");
+            wcscat_s(wzProperName,MAX_FILENAME_LENGTH,W(".il"));
```
Could you please create a github issue to eventually fix the filename size limits in ilasm? We have done similar thing in PAL / CoreCLR, since Unix supports much longer filenames.

krwq commented on src/ilasm/grammar_after.cpp:
```diff
@@ -1500,6 +1501,27 @@ AsmParse::~AsmParse()
+
+    BYTE * p = (BYTE *)lpv;
+
+    // Check for Unicode BOM
```
nit: this is formally called Utf16LE (little endian). Unicode is a general name for character set which codepoints are being encoded

janvorli replied:
@krwq This is a very limited implementation of the Windows API IsTextUnicode. See 
https://msdn.microsoft.com/en-us/library/windows/desktop/dd318672(v=vs.85).aspx
Hence the name.

janvorli commented on src/ilasm/main.cpp:
```diff
@@ -896,11 +916,47 @@ extern "C" int _cdecl wmain(int argc, __in WCHAR **argv)
+        _ASSERTE(length != 0);
 
+        LPWSTR result = new (nothrow) WCHAR[length];
+        _ASSERTE(result != NULL);
```
How about the release build? It would crash if the allocation failed. Could you please change it so that we either print error message and return with an error or use ASSERTE_ALL_BUILDS here? The same probably for the other two asserts.

janvorli commented on src/ilasm/main.cpp:
```diff
@@ -896,11 +916,47 @@ extern "C" int _cdecl wmain(int argc, __in WCHAR **argv)
 {
     return (NULL);
 }
+#endif // !FEATURE_CORECLR
```
Should this be FEATURE_PAL? For Windows, we use the wmain directly, right?

janvorli commented on src/ilasm/writer.cpp:
```diff
@@ -1466,11 +1467,15 @@ HRESULT Assembler::CreatePEFile(__in __nullterminated WCHAR *pwzOutputFilename)
     if(m_wzResourceFile)
     {
+#ifdef FEATURE_PAL
+        report->msg("Warning: The Win32 resource file '%S' is ignored and not emitted on xPlatform.\n");
```
You are missing the parameter for the %S

kyulee1 commented:
Thank you all for comments.
@benpye   I moved IsTextUnicode under palrt.
@kangaroo I set the source property as C++ then applying "-x c++" works. 
@krwq  I create an issue -- https://github.com/dotnet/coreclr/issues/2307. Feel free to jump into to expand it.
@janvorli I incorporated all your feedbacks. Can you take a look again?

janvorli commented on src/ilasm/CMakeLists.txt:
```diff
@@ -19,31 +20,55 @@ set(ILASM_SOURCES
+set_source_files_properties( prebuilt/asmparse.c PROPERTIES LANGUAGE CXX )
+
+if(CLR_CMAKE_PLATFORM_UNIX)
+  add_compile_options(-x c++)
```
I wonder - is the `-x C++` still needed when you've set the asmparse.c language to CXX?

kyulee1 replied:
Yes. That's why I initially converted the extension to cpp even though setting property as CXX. These two are required for unix, or clang fails.

janvorli commented:
LGTM. Thanks!
